traits/ctraits.c

```diff
@@ -4223,34 +4229,60 @@ _trait_delegate(trait_object *trait, PyObject *args)
 |  Sets the appropriate value comparison mode flags of a CTrait instance:
 +----------------------------------------------------------------------------*/
 
-static PyObject *
-_trait_comparison_mode(trait_object *trait, PyObject *args)
+static int
+_set_trait_comparison_mode(trait_object *trait, PyObject *value, void *closure)
 {
-    int comparison_mode;
-
-    if (!PyArg_ParseTuple(args, "i", &comparison_mode)) {
-        return NULL;
+    long comparison_mode = PyLong_AsLong(value);
+    
+    if (comparison_mode == -1 && PyErr_Occurred() != NULL) {
+        return -1;
     }
 
-    trait->flags &= ~(TRAIT_NO_VALUE_TEST | TRAIT_OBJECT_ID_TEST);
     switch (comparison_mode) {
         case 0:
+            trait->flags &= ~TRAIT_COMPARE_MASK;
             trait->flags |= TRAIT_NO_VALUE_TEST;
             break;
         case 1:
+            trait->flags &= ~TRAIT_COMPARE_MASK;
             trait->flags |= TRAIT_OBJECT_ID_TEST;
             break;
         case 2:
+            trait->flags &= ~TRAIT_COMPARE_MASK;
             break;
         default:
-            return PyErr_Format(
+            PyErr_Format(
                 PyExc_ValueError,
-                "The comparison mode must be 0..%d, but %d was specified.",
+                "The comparison mode must be 0..%d, but %ld was specified.",
                 MAXIMUM_COMPARISON_MODE_VALUE, comparison_mode);
+            return -1;
     }
 
-    Py_INCREF(Py_None);
-    return Py_None;
+    return 0;
+}
+
+/*-----------------------------------------------------------------------------
+|  getter for trait comparison mode
++----------------------------------------------------------------------------*/
+
+static PyObject *
+_get_trait_comparison_mode_int(trait_object *trait, void *closure)
+{
+    int i_comparison_mode;
+
+    unsigned int compare_flag = trait->flags & TRAIT_COMPARE_MASK;
+
+    if (compare_flag == TRAIT_NO_VALUE_TEST) {
+        i_comparison_mode = 0;
+    }
+    else if (compare_flag == TRAIT_OBJECT_ID_TEST) {
+        i_comparison_mode = 1;
+    }
+    else {
```

Comment:
In the spirit of defensive programming, it wouldn't be a bad idea either to assert that `compare_flag == TRAIT_EQUALITY_COMPARE` here, or to include the `if (compare_flags == TRAIT_EQUALITY_COMPARE)` in this branch and then add another catch-all `else` branch in which either an assertion is made or `Py_FatalError` is raised.

Reply:
Hmm. `Py_FatalError` wouldn't really be right here. An `assert` would be enough.

Reply:
Yep, `Py_FatalError` is probably not right. I put in an assert.